Ensure armature bones are exported in hierarchical order

timjklein36 opened the pull request:
Fixes #385.

Since some versions of Blender store bones in the order in which they were created, upon export, the bone index in the Godot skeleton might be inconsistent with the actual necessary bone hierarchy to use things like inverse kinematics.

This fixes that by first ordering the bones based on their presence in the armature object tree, respecting real parenting relationships and not just default iteration order. Then the bone index is determined using the existing `generate_bones_mapping` with the supplied _ordered_ bone list.

A test Blend file (`tests/test_scenes/armature/armature_with_arbitrary_bone_creation_order.blend`) for this exact scenario was added, where the bones were created in an intentionally different order to their eventual hierarchical position in the armature.

### _Notes_
- _The `compare` branch of this PR depends on a branch from another open PR, #396, since those changes were necessary for me to accurately test locally._
- _Another reference export needed to be updated, since it also exhibited this inaccurate bone ordering, however it should not affect the end result, since that particular file had no nested bone hierarchy._

scurest commented:
Is the only constraint that parents are ordered before their children? I think `ordered_bones` can be simplified.

Breadth-first traversal is simpler and requires no recursion

````py
def ordered_bones(bones):
    ordered = [bone for bone in bones if bone.parent is None]

    i = 0
    while i < len(ordered):
        ordered += list(ordered[i].children)
        i += 1

    return ordered
````

or depth-first

````py
def ordered_bones(bones):
    ordered = []
    
    def visit(bone):
        nonlocal ordered
        ordered.append(bone)
        for child in bone.children:
            visit(child)

    for bone in bones:
        if bone.parent is None:
            visit(bone)

    return ordered
````

timjklein36 commented:
@scurest Thank you for the suggestions, you are right that it can be simplified. I like the depth-first approach. The results of a depth-first ordering are what I have seen most success with for inverse kinematics, since the bone parenting chains are stored sequentially in Godot.

I also cleaned up the other unrelated changes I was using to test with Docker and squashed the code/reference-export into one commit.

Jason0214 commented:
Thanks very much for fixing it!  The implementation looks very nice : )

I think it does not reply on your docker PR. I will merge this change first.
